### PR TITLE
[OpenAI] Migrate azure-ai-openai to depends on openai-java

### PR DESCRIPTION
**Pull request by mssfang**
Creates a new package `azure-ai-openai-stainless` to hold the changes in `azure-ai-openai`.  Replace `azure-ai-openai` by this new package when `azure-ai-openai` is ready to be shipped as a companion library.

WE SHOULD NOT SHIP `azure-ai-openai-stainless` package

This PR also includes the samples, tests and live test infrastructure for `openai-ai` and our companion library for Azure OpenAI service.

### CODE REVIEW

File: eng/pipelines/templates/steps/build-and-test.yml [hunk 1]
```diff
@@ -142,6 +142,39 @@ steps:
         goals: ${{ parameters.TestGoals }}
       env: ${{ parameters.TestEnvVars }}
 
+
```
Comment (mssfang) [inline eng/pipelines/templates/steps/build-and-test.yml:145]: Need to find a way to not add code into `build-and-test.yml` file which is used by all other SDKs.

**Review by jpalvarezl — APPROVED** [review]
This looks good from my end. Although all the engineering layer stuff, I don't think I have anything of substance to share. Thank you for the work!

File: sdk/openai/azure-ai-openai-stainless/src/main/java/com/azure/ai/openai/stainless/package-info.java
Comment (jpalvarezl) [inline sdk/openai/azure-ai-openai-stainless/src/main/java/com/azure/ai/openai/stainless/package-info.java:1]: Sorry for just commenting here, do we need to delete the source files in under `src/main` ? Also the `tsp-location.yaml` and `assets.json` files?

Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai-stainless/src/main/java/com/azure/ai/openai/stainless/package-info.java:1]: Ok, I just misunderstood that this is a completely new folder. In that case, do you think we should adjust the `README.md` from the `azure-ai-openai` folder?

File: sdk/openai/aisource.yml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        safeName: azureaiopenaistainless
+    CloudConfig:
+      Public:
+        ServiceConnection: azure-sdk-tests-openai
+    EnvVars:
+      AZURE_OPENAI_ENDPOINT: $(AOAI-ENDPOINT-SWEDENCENTRAL)
+      AZURE_OPENAI_KEY: $(AOAI-API-KEY-SWEDENCENTRAL)
+      NON_AZURE_OPENAI_KEY: $(OPENAI-API-KEY)
+      AZURE_TEST_RUN_LIVE: 'true'
+      API_VERSION_GA: '2024-06-01'
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/aisource.yml:20]: I can remove this line as it is not used in current tests configuration. We can always config the EnvVar in this file and get the value by using class Configuration, such as `Configuration.getGlobalConfiguration().get("AZURE_OPENAI_ENDPOINT")`. Currently, the Service API version is harded in the `TestUtils` class

File: sdk/openai/openai-build.yml [hunk 3]
```diff
@@ -0,0 +1,78 @@
+steps:
```
Comment (alzimmermsft) [inline sdk/openai/openai-build.yml:1]: Honestly, this situation is so unique, to me this is a case where we probably should just have a bespoke pipeline.

File: sdk/openai/aisource.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+extends:
+  template: /eng/pipelines/templates/stages/archetype-sdk-tests.yml
+  parameters:
+    ServiceDirectory: openai
+    DisableAzureResourceCreation: true
+    PreSteps:
+      - template: ./../../../../../sdk/openai/openai-build.yml
```
Comment (alzimmermsft) [inline sdk/openai/aisource.yml:7]: I'm pretty sure this template path can be passed where it assumes the root of the directory, ex `/sdk/openai/openai-build.yml`

File: sdk/openai/openai-build.yml (hunk 3 above)
Comment (mssfang) [reply sdk/openai/openai-build.yml:1]: Yeah. it is unique circumstances.

File: eng/pipelines/templates/steps/build-and-test.yml (hunk 1 above)
Comment (mssfang) [reply eng/pipelines/templates/steps/build-and-test.yml:145]: The reason to make changes here is because the Stainless OpenAI SDK allows to build with Java 11+ and can be run with Java 8+. So, the code here is adding a task to run tests with Java 8 additionally ensure SDK has minimum Java version, 8 supported. 

All the other steps including what you described above are covered in another files, openai-build.yml and aisource.yml

Comment (alzimmermsft) [reply eng/pipelines/templates/steps/build-and-test.yml:145]: Yeah, that's what I understood from this.

So, what I'm thinking, rather than mess with a bespoke template, we all logic related to downloading, building, and updating to source versions of `openai-java` to the pre-step script. Then in `aisource.yml` we just remove this matrix filter

```yaml
- JavaTestVersion=^(?!1.8).*
```

At this point, we'll handle the Java 11 build requirements within the build step and since the built project can run on Java 8 we can just run the standard matrix (sans the filtered stuff for From Source, etc).

File: sdk/openai/openai-build.yml (hunk 3 above)
Comment (alzimmermsft) [reply sdk/openai/openai-build.yml:1]: Given the other comment I think I'd rather keep this and have it handle all the setup logic to allow the normal CI matrix be used.

File: sdk/openai/Clone-And-Build.ps1 [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# Define repository details
+$repoUrl = "https://github.com/openai/openai-java.git"
+$branchName = "release-please--branches--main--changes--next"
+
+# Clone the repository and checkout the specified branch
+Write-Host "Cloning repository: $repoUrl"
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/Clone-And-Build.ps1:6]: This file is for build locally test only. It can be removed if needed. Just a convenance for ourself.

File: eng/pipelines/templates/steps/build-and-test.yml (hunk 1 above)
Comment (mssfang) [reply eng/pipelines/templates/steps/build-and-test.yml:145]: Q: That project is expected to work with Java 8 but requires Java 11 to run build, right? If that's true we don't need to do anything special in the test matrix, what we actually need to do is handle all the special logic in the PreStep of the pipeline run.
A: yes, it requires run tests with Java 8, 11, 17, 21 and build tests with 11+.  Yes. the special logic is in the PreSteps of the pipeline line already. But only Run tests under Java 8 is covered here.

File: sdk/openai/openai-build.yml [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  - pwsh: |
```
Comment (alzimmermsft) [inline sdk/openai/openai-build.yml:2]: Use the DevOps `checkout` step rather than rolling our own checkout.

```suggestion
  - checkout: https://github.com/openai/openai-java.git
```

Though I don't know the exact usage of this as we've never checked out other repos.

File: sdk/openai/openai-build.yml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+      # Update the version in the build.gradle.kts file
+      $updatedContent = $buildGradleContent -replace 'version\s*=\s*\"([^\"]+)\"', "version = `"$global:newVersion`""
+      Set-Content -Path $buildGradlePath -Value $updatedContent
+     
+      Write-Host "##vso[task.setvariable variable=NewVersion]$newVersion"
+    displayName: Find current version and update to SNAPSHOT
+    workingDirectory: $(Agent.TempDirectory)
+
+  - pwsh: |
```
Comment (alzimmermsft) [inline sdk/openai/openai-build.yml:49]: Use the [Gradle build task](https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/gradle-v4?view=azure-pipelines) instead as it can manage setting the Java version for you and running the build.

File: eng/pipelines/templates/steps/build-and-test.yml (hunk 1 above)
Comment (alzimmermsft) [reply eng/pipelines/templates/steps/build-and-test.yml:145]: > it requires run tests with Java 8, 11, 17, 21 and build tests with 11+

Right, but we don't need to build tests with Java 11+ as our code here doesn't require Java 11, only building `openai-java` requires 11+ and that's done well before we compile tests.

File: sdk/openai/openai-build.yml (hunk 7 above)
Comment (alzimmermsft) [reply sdk/openai/openai-build.yml:49]: I think this is why we were filtering out the Java 8 runs as we weren't setting a Java version here, leading to Gradle compilation failures.

But since the Gradle task will manage setting the Java version based on `jdkVersionOption` it'll make everything work happily (and we can hardcode to `1.11` as we don't need to verify their project can build with Java 17, 21, etc).

File: sdk/openai/aisource.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      AZURE_OPENAI_ENDPOINT: $(AOAI-ENDPOINT-SWEDENCENTRAL)
+      AZURE_OPENAI_KEY: $(AOAI-API-KEY-SWEDENCENTRAL)
+      NON_AZURE_OPENAI_KEY: $(OPENAI-API-KEY)
+      AZURE_TEST_RUN_LIVE: 'true'
+      API_VERSION_GA: '2024-06-01'
+    MatrixFilters:
+      - JavaTestVersion=^(?!1.8).*
+      - TestFromSource=^$|false
+      - AZURE_TEST_HTTP_CLIENTS=netty
+      - pool=^(?!.*(win)).*
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/aisource.yml:25]: there is no source code but only testing the Stainless' SDK for Azure OpenAI. 
I can remove `AZURE_TEST_HTTP_CLIENTS=netty` as it is not needed in `openai-java`.
`pool=^(?!.*(win)).*` which exclude OS=Windows (IIRC there is some issue with Windows so only run testing with Ubuntu`

File: sdk/openai/openai-build.yml (hunk 7 above)
Comment (mssfang) [reply sdk/openai/openai-build.yml:49]: We probably also need to verify their project can be built with Java 17 and 21 as we add Azure-specific features in their source code.

File: sdk/openai/aisource.yml (hunk 4 above)
Comment (alzimmermsft) [reply sdk/openai/aisource.yml:7]: > do you mean

Basically, except it is `/sdk/openai/openai-build.yml` as template paths lead with just `/`

File: eng/pipelines/templates/steps/build-and-test.yml
Comment (alzimmermsft) [inline eng/pipelines/templates/steps/build-and-test.yml:1]: With the changes to using the `Gradle` task for installing `openai-java`, it appears the changes in the file can be reverted.

File: sdk/openai/aisource.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/openai/aisource.yml:23]: Are the Windows build issues on our side or theirs?

[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/aisource.yml]: image
Since we don't use any http client, can we just limit to use netty and Okhttpclient (Stainless SDK uses OkHttpClient but it doesn't matter in testing our Azure OpenAI feature, as we don't configure which http client to use. We only use what the default http client that the stainless SDK uses)

[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/aisource.yml]: Netty covers 11, 17 and 21 and OkHttpClient covers 8

[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/openai/aisource.yml]: OS=Windows does not find the Gradle Build
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=4758413&view=logs&j=f6513956-70d3-5ac6-ce06-d3a3a7d3344d&t=e8a5492f-ff45-509f-364e-d7b6ddc3e159

image

File: sdk/openai/openai-build.yml (hunk 7 above)
Comment (mssfang) [reply sdk/openai/openai-build.yml]: > We probably also need to verify their project can be built with Java 17 and 21 as we add Azure-specific features in their source code.

Forget what I was saying above. Yeah. if Stainless team already release, they should already compile the source code and run tests successfully for non-azure endpoint.

File: eng/pipelines/templates/steps/build-and-test.yml
Comment (mssfang) [reply eng/pipelines/templates/steps/build-and-test.yml:1]: Will remove changes once the Gradle task works

File: sdk/openai/openai-build.yml (hunk 7 above)
Comment (mssfang) [reply sdk/openai/openai-build.yml:55]: - task: JavaToolInstaller@1
    inputs:
      versionSpec: '11'
      jdkArchitectureOption: 'x64'
      jdkSourceOption: 'PreInstalled'

File: sdk/openai/openai-build.yml (hunk 6 above)
Comment (alzimmermsft) [reply sdk/openai/openai-build.yml:2]: Can deal with this later

**Review by alzimmermsft — APPROVED** [review]
